src/rpc/blockchain.cpp

```diff
@@ -940,6 +940,106 @@ static UniValue getblock(const JSONRPCRequest& request)
+            "  \"merkleroot\" : \"xxxx\", (string) The merkle root\n"
+            "  \"tx\" : [               (array of string) The transaction ids\n"
+            "     \"transactionid\"     (string) The transaction id\n"
+            "     ,...\n"
+            "  ],\n"
+            "  \"time\" : ttt,          (numeric) The block time in seconds since epoch (Jan 1 1970 GMT)\n"
+            "  \"mediantime\" : ttt,    (numeric) The median block time in seconds since epoch (Jan 1 1970 GMT)\n"
+            "  \"nonce\" : n,           (numeric) The nonce\n"
+            "  \"bits\" : \"1d00ffff\", (string) The bits\n"
+            "  \"difficulty\" : x.xxx,  (numeric) The difficulty\n"
+            "  \"chainwork\" : \"xxxx\",  (string) Expected number of hashes required to produce the chain up to this block (in hex)\n"
+            "  \"nTx\" : n,             (numeric) The number of transactions in the block.\n"
+            "  \"previousblockhash\" : \"hash\",  (string) The hash of the previous block\n"
+            "  \"nextblockhash\" : \"hash\"       (string) The hash of the next block\n"
+            "}\n"
+                    },
+                    RPCResult{"for verbosity = 2",
+            "{\n"
+            "  ...,                     Same output as verbosity = 1.\n"
+            "  \"tx\" : [               (array of Objects) The transactions in the format of the getrawtransaction RPC. Different from verbosity = 1 \"tx\" result.\n"
+            "         ,...\n"
+            "  ],\n"
+            "  ,...                     Same output as verbosity = 1.\n"
+            "}\n"
+                    },
+                },
+                RPCExamples{
+                    HelpExampleCli("getblockbyheight", "0")
+            + HelpExampleRpc("getblockbyheight", "42")
+                },
+    };
+
+    if (request.fHelp || !help.IsValidNumArgs(request.params.size())) {
+        throw std::runtime_error(help.ToString());
+    }
+
+    int verbosity = 1;
+    if (!request.params[1].isNull()) {
+        if(request.params[1].isNum())
+            verbosity = request.params[1].get_int();
+        else
+            verbosity = request.params[1].get_bool() ? 1 : 0;
+    }
+
+    CBlock block;
+    // Check if block exists or is negative
+    if(::ChainActive().Height() < request.params[0].get_int() || request.params[0].get_int() <= -1)
+    {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Block not found");
+    }
+    CBlockIndex* pblockindex = ::ChainActive()[request.params[0].get_int()];
+    uint256 hash(ParseHashV(pblockindex->GetBlockHash().GetHex(), "blockhash"));
+    const CBlockIndex* tip;
+    {
+        LOCK(cs_main);
+        pblockindex = LookupBlockIndex(hash);
+        tip = ::ChainActive().Tip();
+
+        if (!pblockindex) {
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Block not found");
+        }
+
+        block = GetBlockChecked(pblockindex);
+    }
+
+    if (verbosity <= 0)
```

Comment:
There is no documentation for when verbosity is less than 0. I don't think it can ever be? Should verbosity be an unsigned int or enum?

Reply:
Nope verbosity cannot be less than zero

Reply:
Can we remove this check for less than zero and change verbosity to an unsigned short or preferably an enum of constants 0-3?

Reply:
I think so

Reply:
IIRC it is undefined behavior to cast an `int` to an `enum` that is out of range of the `enum`. Because any `int` can be entered in the RPC, it could be out of range of the `enum` and cause undesirable things. So I think it is actually better to just handle it as an int. Also, `UniValue` does not have a `get_short()` function so there would be no way to fetch a short from the request other than casting which I think is undesirable especially when it's easier to just handle it as an int.

Reply:
I appreciate the response @achow101.  I would then suggest we add a validate and transform for the RPC call but that is out of the scope of the PR. Validation from the client is necessary in these cases so we don't get undesirable things and it would keep the code clean.

Reply:
Validation from the client is not possible. The client is not necessarily `bitcoin-cli`. It could be any JSON-RPC client. Validation of this must happen server side.

Reply:
> Validation from the client is necessary in these cases so we don't get undesirable things and it would keep the code clean.

No.

Reply:
@fqlx from the server point of view, never trust the client.

Reply:
I would test the argument for valid range and throw an rpc error if out of range. Better to accept just the valid values IMO.